### PR TITLE
add dynamic-loader example

### LINKED ISSUE #3: Add example of dynamic loading and attestation
Add a minimalistic example of how to dynamically load and attest an SM at runtime, using the untrusted [sancus-support](https://github.com/sancus-pma/sancus-support)  host software (ELF loader).

### PR DESCRIPTION
**Pull request by gianlu33**
Fixes #3 

I made a very simple application that loads dynamically an SM at runtime, using `sm_load()`.

The SM is implemented in `sm1.c`. This file is **not** included with the other source files when the application is built, but rather it is built separately and the binary file is automatically hardcoded as a byte array into `sm1.h` using a python script. Later on, this array is passed as input to the `sm_load()` function.

During the loading process, the SM needs information about some symbols (e.g., `exit`, `__unprotected_entry`), hence the `static-symtab.c` file is used to declare such symbols.

The logic of `main.c` is to be decided: @jovanbulck what should this example do/show other than loading dynamically an SM? Just temporarily, this logic checks if the SM's symbols are added correctly to the global symbol table (see sancus-tee/sancus-support#11)

### CODE REVIEW

**Comment by jovanbulck** [comment]
Thanks a lot for the PR @gianlu33 ! This really helps to understand and maintain the dynamic-loading functionality!

I added the example to the GitHub Actions CI as per ccd91cf so we can test and adjust before merging :+1: 

I'll start a code review and I might directly push small changes, or I might follow up with requested changes if it's not clear

File: dynamic-loader/Makefile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+TARGET          = main.elf
+TARGET_NO_WRAP  = no_wrap_$(TARGET)
+
+LDFLAGS += -lsm-control -lloader -ltools
+
+all: $(TARGET)
+
+$(TARGET_NO_WRAP): $(SM1_HEADER) $(OBJECTS)
+	$(LD) $(LDFLAGS) --prepare-for-sm-text-section-wrapping -o $@ $^
```
Comment (jovanbulck) [inline dynamic-loader/Makefile]: better to keep this example minimal without confidential loading. I understand the natural decision to base this on hello-world, but that one is unnecessarily complex w confidential loading, opening a separate issue for this #17

File: dynamic-loader/Makefile [hunk 2]
```diff
@@ -0,0 +1,30 @@
+include ../Makefile.include
+
+SOURCES         = $(shell find *.c ! -name "sm1.c" -printf "%p ")
+OBJECTS         = $(SOURCES:.c=.o)
+SM1_HEADER      = $(shell python3 hardcode_elf.py)
```
Comment (jovanbulck) [inline dynamic-loader/Makefile]: better to use a proper Make target for this as a dependency

current approach also fails in CI:

https://github.com/sancus-tee/sancus-examples/pull/16/checks?check_run_id=3747451902

File: dynamic-loader/hardcode_elf.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import os
+
+# build sm1.elf
+os.system("sancus-cc -c -o sm1.o sm1.c")
```
Comment (jovanbulck) [inline dynamic-loader/hardcode_elf.py]: not sure, but I think this should be possible in the Makefile as well, then we don't have the flags etc hard-coded but can use the ones from Makefile.include

you essentially already did the hard part with having all the logic here neatly in Python so I'll see if I can give it a shot converting it to Makefile syntax :)

File: dynamic-loader/hardcode_elf.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+os.system("sancus-ld --inline-arithmetic -o sm1.elf sm1.o")
+
+# read ELF
+with open("sm1.elf", "rb") as f:
+    sm1 = f.read()
+
+# prepare string to hardcode: remove brackets
```
Comment (jovanbulck) [inline dynamic-loader/hardcode_elf.py]: not sure what's happening here/why this is needed

File: dynamic-loader/hardcode_elf.py (hunk 3 above)
Comment (jovanbulck) [reply dynamic-loader/hardcode_elf.py]: at least the generation of the elf file would be better in Make, we can still keep python to fill in the header I agree

File: dynamic-loader/main.c [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include <sancus_support/global_symtab.h>
+
+#include "sm1.h"
+
+int main()
+{
+    msp430_io_init();
+    printf("Loading SM from buffer..\n");
+    int id = sm_load((void *) SM_BINARY, SM_NAME, VENDOR_ID);
+    printf("Loaded SM id: %d\n", id);
```
Comment (jovanbulck) [inline dynamic-loader/main.c]: nice!

we can consider extending this later to also call the SM to verify it's indeed correctly loaded :thinking:

**Review by jovanbulck — COMMENTED** [review]
Left some comments, will see now if I can start addressing them partially already

**Comment by gianlu33** [comment]
I very much agree that it is better to compile the SM in the Makefile, thanks for making these changes :)

Let me know if you need something else from me!

File: dynamic-loader/main.c (hunk 5 above)
Comment (jovanbulck) [reply dynamic-loader/main.c]: done as per 4f48273

**Comment by jovanbulck** [comment]
This is now merged, I think it's very nice and really helps showing a nice aspect about Sancus! Thanks @gianlu33  for the PR! :+1:

**Pull request merged**